File: rust/src/extract/key_facts.rs
//! Key-facts extractor — top-N fact-bearing sentences as complete strings.
//!
//! Mirrors `src/lede/extract/key_facts.py`. Ranked by composite tfidf + stat
//! density, deduped by normalized (`stat_type`, value), returned in document
//! order. Composition primitive: reuses `stats()` and `composite_score_parts`
//! — no new regex patterns.
//!
//! Rust has no `backend=` argument (only 'regex' emits byte-identical output
//! with Python). The `convert_word_names` flag is surfaced via
//! [`KeyFactsOptions`] mirroring the stats API.

use crate::extract::stats::{StatsOptions, stats_with_options};
use crate::hints::{HintMode, HintWeight, hint_bonus, preprocess_hints, round_to_int};
use crate::sentences::split_sentences;
use crate::tfidf::composite_score_parts;
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::sync::OnceLock;

// Shared with Python's lede/extract/key_facts.py::_STAT_DENSITY_WEIGHT.
const STAT_DENSITY_WEIGHT: f64 = 0.15;
const TFIDF_WEIGHT: f64 = 0.60;
const POSITION_WEIGHT: f64 = 0.25;
const LENGTH_WEIGHT: f64 = 0.15;

/// Candidate row in the ranking loop: (score, sentence index, dedup keys).
type Candidate = (f64, usize, HashSet<(String, String)>);

fn hyphen_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"[-_/]+").expect("static regex"))
}

fn ws_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"\s+").expect("static regex"))
}

/// Normalize a value string for dedup. Matches
/// `benchmarks/extraction_eval.py::_norm` and Python
/// `lede/extract/key_facts.py::_norm_value`.
fn norm_value(s: &str) -> String {
    let lowered = s.to_lowercase();
    let dashed = hyphen_re().replace_all(&lowered, " ");
    ws_re().replace_all(dashed.trim(), " ").to_string()
}

/// Options for [`key_facts_with_options`]. Mirrors Python's keyword-only
/// kwargs on `key_facts()`.
#[derive(Debug, Clone)]
pub struct KeyFactsOptions {
    /// Max number of returned sentences. Default `10`.
    pub max_facts: usize,
    /// Forward to `stats()`; requires the `wordforms` cargo feature when
    /// true. No-op (treated as false) when the feature is off.
    pub convert_word_names: bool,
    /// Hint terms and weights. Empty vec = no-hints path (byte-identical to pre-v0.4).
    pub hints: Vec<HintWeight>,
    /// Fraction of `max_facts` budget reserved for hint-matching sentences.
    /// Range `[0.0, 1.0]`. Default `0.7`.
    pub hint_focus: f64,
    /// Biasing strategy. Default [`HintMode::Soft`].
    pub hint_mode: HintMode,
}

impl Default for KeyFactsOptions {
    fn default() -> Self {
        Self {
            max_facts: 10,
            convert_word_names: false,
            hints: Vec::new(),
            hint_focus: 0.7,
            hint_mode: HintMode::Soft,
        }
    }
}

/// Return top-N sentences containing numeric/named facts.
///
/// See [`key_facts_with_options`] for the full contract.
#[must_use]
pub fn key_facts(text: &str, max_facts: usize) -> Vec<String> {
    key_facts_with_options(
        text,
        KeyFactsOptions {
            max_facts,
            ..Default::default()
        },
    )
}

/// Return top-N sentences containing numeric/named facts.
///
/// Candidates are any sentence that produces at least one `Stat`. Each
/// candidate is scored as:
///
/// ```text
///   0.60*tfidf + 0.25*position + 0.15*length
///   + 0.15 * (num_stats_in_sentence / sentence_len_in_tokens)
/// ```
///
/// Candidates are deduped by normalized (`stat_type`, value): if two
/// sentences share any such key, only the higher-scored one is kept.
/// Ties broken by earlier document position. The top `opts.max_facts`
/// survivors are returned in **document order** (not score order).
///
/// When `opts.hints` is non-empty, the two-pool hint-aware selection
/// mirrors Python T7 exactly, including the hard-mode rollover fix.
#[must_use]
#[allow(
    clippy::cast_precision_loss,
    clippy::too_many_lines,
    clippy::needless_pass_by_value,
    clippy::cast_sign_loss,
    clippy::cast_possible_wrap
)]
pub fn key_facts_with_options(text: &str, opts: KeyFactsOptions) -> Vec<String> {
    if text.is_empty() {
        return Vec::new();
    }
    let all_stats = stats_with_options(
        text,
        StatsOptions {
            convert_word_names: opts.convert_word_names,
        },
    );
    if all_stats.is_empty() {
        return Vec::new();
    }

    let sentences = split_sentences(text);
    if sentences.is_empty() {
        return Vec::new();
    }

    // Build a sentence-string → first-index lookup for O(1) grouping.
    // First occurrence wins — matches Python dict.setdefault behavior and
    // how stats() emits context_sentence.
    let mut sent_to_idx: HashMap<&str, usize> = HashMap::with_capacity(sentences.len());
    for (i, s) in sentences.iter().enumerate() {
        sent_to_idx.entry(s.as_str()).or_insert(i);
    }

    // Group stats by their context_sentence index.
    let mut stats_by_sentence: HashMap<usize, Vec<&crate::extract::stats::Stat>> = HashMap::new();
    for stat in &all_stats {
        if let Some(&idx) = sent_to_idx.get(stat.context_sentence.as_str()) {
            stats_by_sentence.entry(idx).or_default().push(stat);
        }
    }
    if stats_by_sentence.is_empty() {
        return Vec::new();
    }

    let parts = composite_score_parts(&sentences);

    let mut candidates: Vec<Candidate> = Vec::with_capacity(stats_by_sentence.len());
    for (idx, stat_list) in &stats_by_sentence {
        let (t, p, l) = parts[*idx];
        let composite = TFIDF_WEIGHT * t + POSITION_WEIGHT * p + LENGTH_WEIGHT * l;
        let tok_count = sentences[*idx].split_whitespace().count().max(1);
        let density = stat_list.len() as f64 / tok_count as f64;
        let score = composite + STAT_DENSITY_WEIGHT * density;

        let keys: HashSet<(String, String)> = stat_list
            .iter()
            .map(|s| (s.stat_type.clone(), norm_value(&s.value)))
            .collect();
        candidates.push((score, *idx, keys));
    }

    let processed_hints = preprocess_hints(&opts.hints);

    if !processed_hints.is_empty() {
        // Two-pool hint-aware selection — mirrors Python T7 exactly.
        // Compute per-candidate hint bonuses.
        let bonus_for: Vec<f64> = candidates
            .iter()
            .map(|(_score, idx, _keys)| hint_bonus(&sentences[*idx], &processed_hints))
            .collect();

        // hint_keyed: (sort_key, doc_idx, cand_idx)
        // hard mode sends bonus==0 to the bottom with f64::INFINITY sentinel.
        let mut hint_keyed: Vec<(f64, usize, usize)> = candidates
            .iter()
            .enumerate()
            .map(|(i, (score, idx, _))| {
                let bonus = bonus_for[i];
                let sort_key = if opts.hint_mode == HintMode::Hard && bonus == 0.0 {
                    f64::INFINITY
                } else {
                    -(score + bonus)
                };
                (sort_key, *idx, i)
            })
            .collect();
        hint_keyed.sort_by(|a, b| {
            a.0.partial_cmp(&b.0)
                .unwrap_or(std::cmp::Ordering::Equal)
                .then(a.1.cmp(&b.1))
        });

        let mut plain_keyed: Vec<(f64, usize, usize)> = candidates
            .iter()
            .enumerate()
            .map(|(i, (score, idx, _))| (-score, *idx, i))
            .collect();
        plain_keyed.sort_by(|a, b| {
            a.0.partial_cmp(&b.0)
                .unwrap_or(std::cmp::Ordering::Equal)
                .then(a.1.cmp(&b.1))
        });

        let hint_quota = round_to_int(opts.max_facts as i64, opts.hint_focus) as usize;
        let normal_quota = opts.max_facts - hint_quota;

        let mut kept_keys: HashSet<(String, String)> = HashSet::new();
        let mut kept: Vec<(f64, usize)> = Vec::new();
        let mut selected_cand_indices: HashSet<usize> = HashSet::new();

        // Returns true if the candidate passes dedup and is kept.
        // Uses a closure-style helper via a macro pattern since closures
        // can't borrow `candidates` and `kept_keys`/`kept`/`selected` simultaneously.
        macro_rules! try_keep {
            ($cand_idx:expr) => {{
                let (c_score, c_doc_idx, ref c_keys) = candidates[$cand_idx];
                if c_keys.iter().any(|k| kept_keys.contains(k)) {
                    false
                } else {
                    kept_keys.extend(c_keys.iter().cloned());
                    kept.push((c_score, c_doc_idx));
                    selected_cand_indices.insert($cand_idx);
                    true
                }
            }};
        }

        // Hint pool: walk hint_keyed until hint_quota deduped sentences kept.
        let mut hint_kept_count: usize = 0;
        for &(sort_key, _doc, cand_idx) in &hint_keyed {
            if opts.hint_mode == HintMode::Hard && sort_key == f64::INFINITY {
                continue; // bonus==0 in hard mode — ineligible
            }
            if selected_cand_indices.contains(&cand_idx) {
                continue;
            }
            if hint_kept_count >= hint_quota {
                break;
            }
            if try_keep!(cand_idx) {
                hint_kept_count += 1;
            }
        }

        // Plain pool: walk plain_keyed until normal_quota MORE kept.
        let mut normal_kept_count: usize = 0;
        for &(_neg_score, _doc, cand_idx) in &plain_keyed {
            if selected_cand_indices.contains(&cand_idx) {
                continue;
            }
            if normal_kept_count >= normal_quota {
                break;
            }
            if try_keep!(cand_idx) {
                normal_kept_count += 1;
            }
        }

        // Rollover: hint pool came up short → take more.
        // Hard mode: stay restricted to hint-bearing candidates (hint_keyed,
        // skipping INFINITY-sentinel entries). Soft mode: draw from plain_keyed.
        if hint_kept_count < hint_quota {
            let extra_needed_start = hint_quota - hint_kept_count;
            let mut extra_needed = extra_needed_start;
            if opts.hint_mode == HintMode::Hard {
                for &(sort_key, _doc, cand_idx) in &hint_keyed {
                    if sort_key == f64::INFINITY {
                        continue; // non-matching in hard mode
                    }
                    if selected_cand_indices.contains(&cand_idx) {
                        continue;
                    }
                    if extra_needed == 0 {
                        break;
                    }
                    if try_keep!(cand_idx) {
                        extra_needed -= 1;
                    }
                }
            } else {
                for &(_neg_score, _doc, cand_idx) in &plain_keyed {
                    if selected_cand_indices.contains(&cand_idx) {
                        continue;
                    }
                    if extra_needed == 0 {
                        break;
                    }
                    if try_keep!(cand_idx) {
                        extra_needed -= 1;
                    }
                }
            }
        }

        // Soft-only rollover: plain pool came up short → hint pool takes more.
        if opts.hint_mode == HintMode::Soft && normal_kept_count < normal_quota {
            let mut extra_needed = normal_quota - normal_kept_count;
            for &(sort_key, _doc, cand_idx) in &hint_keyed {
                let _ = sort_key;
                if selected_cand_indices.contains(&cand_idx) {
                    continue;
                }
                if extra_needed == 0 {
                    break;
                }
                if try_keep!(cand_idx) {
                    extra_needed -= 1;
                }
            }
        }

        // Return in document order.
        kept.sort_by_key(|c| c.1);
        return kept
            .into_iter()
            .map(|(_, i)| sentences[i].clone())
            .collect();
    }

    // No hints: original single-pool dedup path (byte-identical to pre-v0.4).
    // Sort score-descending, idx-ascending (deterministic tie-break).
    candidates.sort_by(|a, b| {
        b.0.partial_cmp(&a.0)
            .expect("no NaN in scores")
            .then_with(|| a.1.cmp(&b.1))
    });

    let mut kept_keys: HashSet<(String, String)> = HashSet::new();
    let mut kept: Vec<usize> = Vec::with_capacity(opts.max_facts);
    for (_, idx, keys) in candidates {
        if keys.iter().any(|k| kept_keys.contains(k)) {
            continue;
        }
        kept_keys.extend(keys);
        kept.push(idx);
        if kept.len() >= opts.max_facts {
            break;
        }
    }

    // Return in document order.
    kept.sort_unstable();
    kept.into_iter().map(|i| sentences[i].clone()).collect()
}

File: src/lede/extract/key_facts.py
"""Key-facts extractor — top-N fact-bearing sentences as complete strings.

Ranked by composite tfidf + stat density, deduped by normalized
(stat_type, value), returned in document order. Composition primitive:
reuses ``stats()`` and ``_composite_score_parts`` — no new regex patterns.

Mirrors rust/src/extract/key_facts.rs. Registered as ('regex','key_facts')
for registry symmetry with the other primitives. No spaCy backend (spaCy
composition path would be a different primitive, not a different backend).
"""
from __future__ import annotations

import re

from ..sentences import split_sentences
from ..tfidf import _composite_score_parts, _TFIDF_WEIGHT, _POSITION_WEIGHT, _LENGTH_WEIGHT
from .._hints import preprocess_hints, hint_bonus, round_to_int
from ._backends import register
from .stats import stats as _stats

# Normalization rule shared with benchmarks/extraction_eval.py::_norm — kept
# in sync by convention (same semantics, local copy to avoid bench dep).
_HYPHEN_RE = re.compile(r"[-_/]+")


def _norm_value(s: str) -> str:
    """Lowercase, hyphen/underscore/slash → space, collapse whitespace.

    Used for dedup key on stat values so "five-day" and "five day" are
    considered the same fact.
    """
    return re.sub(r"\s+", " ", _HYPHEN_RE.sub(" ", s.lower())).strip()


# Stat-density contribution weight in the final score. Kept small so the
# composite tfidf signal still dominates.
_STAT_DENSITY_WEIGHT = 0.15


def key_facts(
    text: str,
    *,
    max_facts: int = 10,
    convert_word_names: bool = False,
    hints: list[str] | dict[str, float] | None = None,
    hint_focus: float = 0.7,
    hint_mode: str = "soft",
) -> tuple[str, ...]:
    """Return top-N sentences containing numeric/named facts.

    Candidates are any sentence that produces at least one ``Stat`` from
    ``stats(text, convert_word_names=convert_word_names)``. Each candidate
    is scored as:

        0.60*tfidf + 0.25*position + 0.15*length
        + 0.15 * (num_stats_in_sentence / sentence_len_in_tokens)

    Candidates are deduped by the normalized ``(stat_type, value)`` key:
    if two sentences share any such key, only the higher-scored one is
    kept. Ties broken by earlier document position (deterministic).

    The top ``max_facts`` survivors are returned in **document order**
    (not score order) — readers want reading flow.

    Args:
        text: input document text.
        max_facts: cap on number of returned sentences. Default 10.
        convert_word_names: forwarded to ``stats()``; requires the
            ``[wordforms]`` extra when True.
        hints: optional list[str] or dict[str, float] of terms to bias
            selection toward. List entries get weight 1.0. Default None.
        hint_focus: fraction of ``max_facts`` quota reserved for
            hint-matching candidates. Range [0.0, 1.0]. Default 0.7.
            Ignored when ``hints`` is None.
        hint_mode: ``"soft"`` (default) or ``"hard"``. ``"soft"`` reorders
            hint-bearing candidates to fill the hint quota first; ``"hard"``
            restricts the hint pool to candidates that contain at least one
            hint. Ignored when ``hints`` is None.

    Returns:
        Tuple of complete sentence strings (not ``Stat`` tuples). Empty
        tuple when no sentence contains a recognized fact.

    Raises:
        ValueError: on ``hint_focus`` outside [0.0, 1.0] or unknown
            ``hint_mode``.

    Hint biasing (v0.4):
        When ``hints`` is None (the default), no new code path executes and
        output is byte-identical to v0.3.0. The two-pool selection uses
        ``max_facts`` as the count budget (not chars): ``hint_quota =
        round(max_facts * hint_focus)``, ``normal_quota = max_facts -
        hint_quota``. Deduplication by ``(stat_type, normalized_value)``
        runs across both pools combined — a high-scoring hint-bearing
        sentence can block a duplicate from appearing in the plain pool.

        Matching is case-insensitive and word-boundary-delimited (``\\b``);
        no Unicode normalization or stemming. See docs/REFERENCE.md "Hint
        biasing" for the full contract.
    """
    if not text:
        return ()

    processed_hints = preprocess_hints(hints)
    if processed_hints:
        if not 0.0 <= hint_focus <= 1.0:
            raise ValueError(f"hint_focus must be in [0.0, 1.0]; got {hint_focus}")
        if hint_mode not in ("soft", "hard"):
            raise ValueError(f"hint_mode must be 'soft' or 'hard'; got {hint_mode!r}")

    all_stats = _stats(text, convert_word_names=convert_word_names)
    if not all_stats:
        return ()

    sentences = split_sentences(text)
    if not sentences:
        return ()

    # Group stats by their context_sentence. We walk sentences in order and
    # match on string equality (context_sentence is a direct slice of the
    # input, same as the one split_sentences returns).
    stats_by_sentence: dict[int, list] = {}
    # Build a sentence-string → index lookup for O(1) grouping. When the same
    # sentence appears twice in a document the first occurrence wins — this
    # is deterministic and matches how stats() emits context_sentence.
    sent_to_idx: dict[str, int] = {}
    for i, s in enumerate(sentences):
        sent_to_idx.setdefault(s, i)
    for stat in all_stats:
        idx = sent_to_idx.get(stat.context_sentence)
        if idx is None:
            continue
        stats_by_sentence.setdefault(idx, []).append(stat)

    if not stats_by_sentence:
        return ()

    parts = _composite_score_parts(sentences)

    candidates: list[tuple[float, int, list]] = []
    for idx, stat_list in stats_by_sentence.items():
        t, p, l_ = parts[idx]
        composite = _TFIDF_WEIGHT * t + _POSITION_WEIGHT * p + _LENGTH_WEIGHT * l_
        # Sentence length in whitespace-separated tokens; mirrors how
        # length_score counts words. Guarded against zero for safety.
        tok_count = max(1, len(sentences[idx].split()))
        density = len(stat_list) / tok_count
        score = composite + _STAT_DENSITY_WEIGHT * density
        candidates.append((score, idx, stat_list))

    if processed_hints:
        # Two-pool hint-aware selection.
        # Compute per-candidate hint bonuses.
        bonus_for: list[float] = [
            hint_bonus(sentences[idx], processed_hints)
            for _score, idx, _stat_list in candidates
        ]

        # Build sorted orderings for each pool.
        # hint_keyed: rank by (score + bonus); hard mode sends bonus==0 to bottom.
        hint_keyed: list[tuple[float, int, int]] = []
        for i, ((score, idx, _stat_list), bonus) in enumerate(
            zip(candidates, bonus_for)
        ):
            if hint_mode == "hard" and bonus == 0:
                sort_key = float("inf")
            else:
                sort_key = -(score + bonus)
            hint_keyed.append((sort_key, idx, i))
        hint_keyed.sort()

        plain_keyed: list[tuple[float, int, int]] = sorted(
            [(-c[0], c[1], i) for i, c in enumerate(candidates)]
        )

        hint_quota = round_to_int(max_facts, hint_focus)
        normal_quota = max_facts - hint_quota

        kept_keys: set[tuple[str, str]] = set()
        kept: list[tuple[float, int]] = []
        selected_cand_indices: set[int] = set()

        def _try_keep(cand_idx: int) -> bool:
            """Keep candidate if it passes the dedup filter. Returns True on success."""
            c_score, c_doc_idx, c_stat_list = candidates[cand_idx]
            keys = {(s.stat_type, _norm_value(s.value)) for s in c_stat_list}
            if keys & kept_keys:
                return False
            kept_keys.update(keys)
            kept.append((c_score, c_doc_idx))
            selected_cand_indices.add(cand_idx)
            return True

        # Hint pool: walk hint_keyed until hint_quota fact-deduped sentences kept.
        hint_kept_count = 0
        for sort_key, _doc, cand_idx in hint_keyed:
            if hint_mode == "hard" and sort_key == float("inf"):
                # bonus == 0 in hard mode — ineligible for hint pool.
                continue
            if cand_idx in selected_cand_indices:
                continue
            if hint_kept_count >= hint_quota:
                break
            if _try_keep(cand_idx):
                hint_kept_count += 1

        # Plain pool: walk plain_keyed until normal_quota MORE kept.
        normal_kept_count = 0
        for _neg_score, _doc, cand_idx in plain_keyed:
            if cand_idx in selected_cand_indices:
                continue
            if normal_kept_count >= normal_quota:
                break
            if _try_keep(cand_idx):
                normal_kept_count += 1

        # Rollover: hint pool came up short → take more.
        # In hard mode, stay restricted to hint-bearing candidates (walk hint_keyed,
        # skipping the float("inf") sentinel entries that mark non-matching items).
        # In soft mode, draw from plain_keyed (any candidate).
        if hint_kept_count < hint_quota:
            extra_needed = hint_quota - hint_kept_count
            rollover_pool = hint_keyed if hint_mode == "hard" else plain_keyed
            for sort_key, _doc, cand_idx in rollover_pool:
                if hint_mode == "hard" and sort_key == float("inf"):
                    # non-matching candidate in hard-mode hint pool — skip
                    continue
                if cand_idx in selected_cand_indices:
                    continue
                if extra_needed <= 0:
                    break
                if _try_keep(cand_idx):
                    extra_needed -= 1

        # Soft-only rollover: plain pool came up short → hint pool takes more.
        if hint_mode == "soft" and normal_kept_count < normal_quota:
            extra_needed = normal_quota - normal_kept_count
            for sort_key, _doc, cand_idx in hint_keyed:
                if cand_idx in selected_cand_indices:
                    continue
                if extra_needed <= 0:
                    break
                if _try_keep(cand_idx):
                    extra_needed -= 1

        # Return in document order.
        kept.sort(key=lambda c: c[1])
        return tuple(sentences[i] for _, i in kept)

    # No hints: original single-pool dedup path (byte-identical to pre-T7).
    # Dedup by (stat_type, normalized value). Walk candidates in
    # score-descending order; for each, if any of its (stat_type, norm_value)
    # keys collide with an already-kept candidate, drop the later one.
    # Ties on score broken by earlier document position.
    candidates.sort(key=lambda c: (-c[0], c[1]))
    kept_keys_sp: set[tuple[str, str]] = set()
    kept_sp: list[tuple[float, int]] = []
    for score, idx, stat_list in candidates:
        keys = {(s.stat_type, _norm_value(s.value)) for s in stat_list}
        if keys & kept_keys_sp:
            continue
        kept_keys_sp.update(keys)
        kept_sp.append((score, idx))
        if len(kept_sp) >= max_facts:
            break

    # Return in document order.
    kept_sp.sort(key=lambda c: c[1])
    return tuple(sentences[i] for _, i in kept_sp)


register("regex", "key_facts", key_facts)
